x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java

```diff
@@ -129,6 +131,43 @@ public void testFailLeaderReplicaShard() throws Exception {
         }
     }
 
+    public void testChangeHistoryUUID() throws Exception {
+        try (ReplicationGroup leaderGroup = createGroup(0);
+             ReplicationGroup followerGroup = createFollowGroup(0)) {
+            leaderGroup.startAll();
+            int docCount = leaderGroup.appendDocs(randomInt(64));
+            leaderGroup.assertAllEqual(docCount);
+            followerGroup.startAll();
+            ShardFollowNodeTask shardFollowTask = createShardFollowTask(leaderGroup, followerGroup);
+            final SeqNoStats leaderSeqNoStats = leaderGroup.getPrimary().seqNoStats();
+            final SeqNoStats followerSeqNoStats = followerGroup.getPrimary().seqNoStats();
+            shardFollowTask.start(
+                leaderSeqNoStats.getGlobalCheckpoint(),
+                leaderSeqNoStats.getMaxSeqNo(),
+                followerSeqNoStats.getGlobalCheckpoint(),
+                followerSeqNoStats.getMaxSeqNo());
+            leaderGroup.syncGlobalCheckpoint();
+            leaderGroup.assertAllEqual(docCount);
+            Set<String> indexedDocIds = getShardDocUIDs(leaderGroup.getPrimary());
+            assertBusy(() -> {
+                assertThat(followerGroup.getPrimary().getGlobalCheckpoint(), equalTo(leaderGroup.getPrimary().getGlobalCheckpoint()));
+                followerGroup.assertAllEqual(indexedDocIds.size());
+            });
+
+            String oldHistoryUUID = leaderGroup.getPrimary().getHistoryUUID();
+            leaderGroup.reinitPrimaryShard();
+            leaderGroup.getPrimary().store().bootstrapNewHistory();
+            recoverShardFromStore(leaderGroup.getPrimary());
+            String newHistoryUUID = leaderGroup.getPrimary().getHistoryUUID();
+
+            assertBusy(() -> {
+                assertThat(shardFollowTask.isStopped(), is(true));
+                assertThat(shardFollowTask.getFailure().getMessage(), equalTo("unexpected history uuid, expected [" + oldHistoryUUID +
```

Comment:
Not sure why this test fails on my machine.
```
java.lang.AssertionError: 
Expected: "unexpected history uuid, expected [_x3AmMhzQsa91ee-ah9Aug], actual [ZWUMypAWSEq9P3BEFIAGfA]"
     but: was "retrying failed [11] times, aborting..."
```

Reply:
I think that is because it took to long to re-initialise the leader primary shard. Currently we retry up to 10 times and then fail. This will change when I merge #33371, which I will do before merging this PR and then this failure will not occur.